### ISSUE TITLE
Saturated vets-api server pod solution

### ISSUE THREAD

**Issue by LindseySaari** [department-of-veterans-affairs/va.gov-team#49794]
## Description
When vets-api-servers are backlogged, are they taken out of rotation or still taking requests? Can they be flagged as "NotReady" to stop them from being further backlogged?

Additionally, are the current puma backlog metrics **only** accounting for those that are still serving traffic -- excluding the backlog count of those no longer taking traffic?

-----

**The current setup allows requests to hit pods that already have a backlog, and this solves that.
what the readiness-probe CM does:**
- runs [locally] in each pod
- hits the /stats endpoint and sums up the backlog
- the script exits using the backlog count (anything > 0 is considered an error)
- a failed readiness probe = keep the pod alive so that it can process the work it has, but don't send any more traffic to it
- the probe is executed every 5 seconds
- if the probe remains failed (backlog > 0) for 15 minutes the pod is terminated -- at this duration, i figure something has gone terribly wrong
- if readiness is already in a failed state but then succeeds (backlog = 0) two times in a row, it's marked as ready and begins receiving traffic
- i figure that a backlog of 0 should succeed twice before considering the pod ready (to make sure the pod has time to settle)

```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: {{ .Release.Namespace }}
  creationTimestamp: null
  name: readiness-probe
data:
  readiness.sh: |
    #!/bin/bash

    # Uses sum of backlog as the exit code
    #
    # If backlog is > 0 then the exit code indicates a failure and the readiness
    #   probe will fail. Once the backlog is back down to 0, the script will
    #   exit with a success and the pod will be placed back into rotation to
    #   accept traffic.

    exit $(
      curl -s 127.0.0.1:9293/stats \
      | grep -oE 'backlog":[0-9]*' \
      | cut -d':' -f2 \
      | awk -v SUM="0" '{ SUM += $1} END { print SUM }'
    )
```

**Adding to the deployment (image):**
1. Add the readiness-probe configmap as a volume to the deployment
2. this will be for the vets-api-web container
3. update the readiness probe settings for the vets-api-web container
mount the script to the container

![image (9).png](https://images.zenhubusercontent.com/5d4836dda95c51615db9bdc9/6c6d361d-32a2-4410-8368-762281e6087d)

## Tasks
- [ ] Investigate how saturated pods receive/don't receive requests
- [ ] Investigate metrics

**Comment by considerable** [comment]
K8s documentation says,
Sometimes, applications are temporarily unable to serve traffic. ...In such cases, you don't want to kill the application, but you don't want to send it requests either. Kubernetes provides readiness probes to detect and mitigate these situations. 
https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes

In other words, if vets-api-servers themselves have a way to know they are backlogged, we could program a readiness endpoint to tell K8s when to exclude a particular vets-api-servers pod from the request receiving rotation.

**Comment by laineymajor** [comment]
Can wait until complete with vets-api server migration in dev to EKS

**Comment by laineymajor** [comment]
@LindseySaari to schedule time to collab with kyle

**Comment by laineymajor** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Platform-Tech-Team-1-633af4074573d06c3cda142a/issues/department-of-veterans-affairs/va.gov-team/49794?planning-poker) with Zenhub @LindseySaari @RachalCassity

**Comment by laineymajor** [comment]
@oseasmoran73 please review and let us know if this is a ticket you'd like to jump in and assist

**Comment by LindseySaari** [comment]
I think this needs to actually be a liveness probe. Readiness runs on startup and Liveness will run on an interval after the pod is already running

**Comment by laineymajor** [comment]
@LindseySaari and @RachalCassity to work with @oseasmoran73  on this... get him up to speed.

There should be a threshold that you can apply to a live readiness probe

**Comment by oseasmoran73** [comment]
Threshold can be found [here](https://www.containiq.com/post/kubernetes-liveness-probe)

**Comment by LindseySaari** [comment]
I was mistaken. This should be a readiness probe. 

Liveness Probe—indicates if the container is operating. If so, no action is taken. If not, the kubelet kills and restarts the container. (we don't want the pod killed, we just want it to stop taking on traffic if saturated) 

Readiness Probe—indicates whether the application running in the container is ready to accept requests. If so, Services matching the pod are allowed to send traffic to it. If not, the endpoints controller removes the pod from all matching Kubernetes Services. (what we want if things are saturated. We want the endpoints controller to remove it from rotation, not completely kill it)

I plan to start working on this next as time permits

**Comment by LindseySaari** [comment]
[PR here for part 1](https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/1657). I plan on doing this in two parts:

1) Add the readiness.sh configmap, volume and volumeMount
2) Invoke the readiness probe
    - Edit the current readiness probe to call the script
    - see if we should move the current readiness probe to a startup probe

**Comment by LindseySaari** [comment]
Added the probe and new script to dev: https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/1659

Todo:
1. Add to higher envs
2. Consider adding the previous readinessProbe as a startUp probe (needs discovery)

**Comment by LindseySaari** [comment]
Merged and works in dev!

[PR for higher envs here](https://github.com/department-of-veterans-affairs/vsp-infra-application-manifests/pull/1660)

Todo:
1. Consider adding the previous readinessProbe as a startUp probe (needs discovery)

**Comment by RachalCassity** [comment]
@LindseySaari Reading these [docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).

Caution: Liveness probes do not wait for readiness probes to succeed. If you want to wait before executing a liveness probe you should use initialDelaySeconds or a startupProbe.

**Comment by RachalCassity** [comment]
From the docs, livenessProbes rely on startupProbes, not readinessProbes.

**Comment by RachalCassity** [comment]
Other notes:

https://elastisys.com/the-difference-between-liveness-and-readiness-probes-in-kubernetes/

**Comment by LindseySaari** [comment]
Todo:
1. startupProbe
2. Change the TCP liveness probe

cc @RachalCassity

**Comment by RachalCassity** [comment]
The initial delay may need to be different between the livenessProbe and startupProbe

**Comment by LindseySaari** [comment]
A solution is in place for the saturated pods via readiness probe. We had some clashes between readiness and liveness yesterday, but this is deemed complete.
